### PR TITLE
Initial plugins code cleanup

### PR DESCRIPTION
**Pull request by olliestanley**
The primary contributions of this PR are:

- Introduce `PromptedLLM` class to manage the repeated process of preparing prompt with memory and template, calling LLM, and postprocessing output
- Separating out functions in `openapi_parser`
- General improvements to clarity/readability

### CODE REVIEW

**Review by AbdBarho — COMMENTED** [review]
I feel abit unqualified to review the internals of the plugin system, but overall it looks good.

File: inference/worker/chat_chain_utils.py [hunk 1]
```diff
@@ -153,28 +152,39 @@ def prepare_json(json_str: str) -> str:
...
-        ((tool, similarity(tool.name, tool_name)) for tool in tools), key=lambda x: x[1], default=(None, 0)
+def select_tool(tool_name: str, tools: list[Tool]) -> Tool | None:
+    tool = next((t for t in tools if t.name in tool_name), None)
+    if tool:
+        return tool
+    tool, tool_similarity = max(
+        ((t, similarity(t.name, tool_name)) for t in tools),
+        key=lambda x: x[1],
+        default=(None, 0),
     )
```
Comment (draganjovanovich) [inline inference/worker/chat_chain_utils.py:171]: I would leave this comment somewhere if possible, just so we do not have a better model in the future, and similarity like it's still used, etc...
# This should become stricter and stricter as we get better models

Comment (olliestanley) [reply inference/worker/chat_chain_utils.py:171]: Ok, added

File: inference/worker/chat_chain_utils.py [hunk 2]
```diff
@@ -293,34 +296,27 @@ def func(req):
...
             description=f"{openapi_specification_title}{parameters_description}{payload_description}tool description: {endpoint.summary}\n",
         )
         tools.append(tool)
 
     tools_string = "\n".join([f"> {tool.name}{tool.description}" for tool in tools])
-    # NOTE: This can be super long for some plugins, that I tested so far.
-    # and because we don't have 32k CTX size, we need to truncate it.
-    plugin_description_for_model = truncate_str(llm_plugin["description_for_model"], 512)
+    # This can be long for some plugins, we need to truncate due to ctx limitations
+    plugin_description_for_model = truncate_str(llm_plugin.description_for_model, 512)
```
Comment (draganjovanovich) [inline inference/worker/chat_chain_utils.py:314]: we could extract 512 to some constant at the top of the file(DESCRIPTION_FOR_MODEL_MAX_LENGTH = 256 ), maybe even reduce it to 256 for now, I forgot that part on my end. sorry...

**Comment by draganjovanovich** [comment]
Nice job! 
LGTM, 
But I would love to manually test this branch a bit more before merging if possible?

File: inference/worker/chat_chain_utils.py (hunk 2 above)
Comment (olliestanley) [reply inference/worker/chat_chain_utils.py]: Makes sense, done

**Comment by olliestanley** [comment]
> Nice job! LGTM, But I would love to manually test this branch a bit more before merging if possible?

Sure, no problem. Let me know when you're happy for it to be merged

**Comment by olliestanley** [comment]
@draganjovanovich the last commit I have pushed to this branch should address #3146. Please let me know if you think there is any problem with this solution.

**Comment by draganjovanovich** [comment]
> @draganjovanovich the last commit I have pushed to this branch should address #3146. Please let me know if you think there is any problem with this solution.

yes, yes. I had that in my local repo, but left it uncommitted. Thanks.

**Comment by draganjovanovich** [comment]
> 

I tested it on a local setup, and everything seems fine.

**Comment by draganjovanovich** [comment]
I have some more improvements for plugins that would like to open PR, but It would be best if that could be merged before that?

**Review by yk — APPROVED** [review]

**Pull request merged**